File: assets/rstudio/customer-attrition-prediction-analytics-dashboard/clientPanel.R
# Sample Materials, provided under license.
# Licensed Materials - Property of IBM
# © Copyright IBM Corp. 2019, 2020. All Rights Reserved.
# US Government Users Restricted Rights - Use, duplication or disclosure restricted by GSA ADP Schedule Contract with IBM Corp.



clientPanel <- function() {
  
  tabPanel(
    "Client View",
    value = "clientPanel",
    
    panel(
      br(),br(),
      fluidRow(
        column(3, class = "pull-left",
               div(id = "customerImage"),
               h2(class = "text-center" ,textOutput("customerName"))),
        column(4, 
               h4("Personal Information", class="text-center"),
               hr(),
               uiOutput("customerInfo")
        ),
        column(4, class = "pull-right", 
               h4("Financial Profile", class="text-center"),
               hr(),
               uiOutput("customerFinancesInfo")
        )
      ),
      br()
    ),
    
    panel(
      h3("Account Details"),
      br(),
      column(12, offset = 1, tableOutput("customerAccountsTable")),
      br(),
      column(12, offset = 1,tableOutput("customerAccountSummaryTable"))
    ),
    
    panel(
      h3("Attrition Prediction "),
      br(),
      

      
      tags$div(
        id = "authPanel",
        column(4,
          panel(
            h4("Connect to Cloud Pak for Data API"),
            textInput("hostname", "CPD Hostname"),
            textInput("username", "CPD Username"),
            passwordInput("password", "CPD Password"),
            actionButton("authBtn", "Authenticate API", class = "btn-primary btn-lg btn-block", style = "max-width:300px", disabled = TRUE),
            tags$head(tags$style("#authError{color:red;}")),
            verbatimTextOutput("authError")
          ),
          style = "max-width:360px;"
        )
      ),
      hidden(
        tags$div(
          id = "deploymentPanel",
          column(4,
             panel(
               tags$h4("Model Scoring Pipeline Deployment"),
               pickerInput(
                 inputId = 'deploymentSelector',
                 label = 'Deployment:',
                 choices = list(),
                 options = pickerOptions(width = "auto", style = "btn-primary")
               ),
               tags$p(
                 tags$strong("Space Name: "),
                 textOutput(outputId = "space_name", inline = TRUE)
               ),
               tags$p(
                 tags$strong("GUID: "),
                 textOutput(outputId = "deployment_guid", inline = TRUE)
               ),
               #tags$p(
               #  tags$strong("Tags: "),
               #  textOutput(outputId = "deployment_tags", inline = TRUE),
               #  style = "word-wrap: break-word"
               #),
               tags$p(
                 tags$strong("Scoring Endpoint: "),
                 textOutput(outputId = "scoring_url", inline = TRUE),
                 style = "word-wrap: break-word"
               )
             ),
             panel(
               actionButton(
                 "reauthenticateBtn",
                 "Re-Authenticate",
                 class = "btn-primary btn-lg btn-block"
               ) 
             )
             
          ),
          tags$div(id = "scoreBtnSection",
            column(4,
              br(),br(),
              actionButton(
                 "scoreBtn",
                 "Predict Attrition",
                 class = "btn-primary btn-lg btn-block",
                 disabled = TRUE
               ),
              br(),
              h4("Input JSON:"),
              verbatimTextOutput("pipelineInput"),
              br(),
              tags$head(tags$style("#scoringError{color:red;}")),
              verbatimTextOutput("scoringError"))
          ),
          column(8,
             hidden(
               tags$div(id = "scoringResponse")
             )
          )
        )
      )
    )
  )
}

# Reactive server variables store (pervades across all sessions)
serverVariables = reactiveValues(deployments = list(), token = '')

if(nchar(Sys.getenv('CP4D_HOSTNAME')) > 0 && nchar(Sys.getenv('CP4D_USERNAME')) > 0 && nchar(Sys.getenv('CP4D_PASSWORD')) > 0) {
  tryCatch({
    deploymentsResp = collectDeployments(Sys.getenv('CP4D_HOSTNAME'), Sys.getenv('CP4D_USERNAME'), Sys.getenv('CP4D_PASSWORD'), "Customer-Attrition-Prediction-Scoring-Function-Deployment")
    serverVariables$deployments <- deploymentsResp$deployments
    serverVariables$token = deploymentsResp$token
  }, warning = function(w) {
    print(w$message)
  }, error = function(e) {
    print(e$message)
  })
}

clientServer <- function(input, output, session, sessionVars) {
  
  observe({

    client <- clients[[toString(sessionVars$selectedClientId)]]

    # Update client name & image
    output$customerName <- renderText(client$name)
    removeUI(selector = "#customerImage > *")
    insertUI(
      selector = "#customerImage",
      where = "beforeEnd",
      ui = img(src = paste0("profiles/",client$image), style = "display: block;margin-left: auto;margin-right: auto;", width=150, height=150)
    )
    
    # Load customer data for customer sessionVars$selectedClientId
    selection <- customer[customer$CUSTOMER_ID == sessionVars$selectedClientId,][1,]
    
    # Table displays for Customer View
    # output$customerTable <- renderTable({
    #   demoDeets <- selection[,c("CUSTOMER_ID", "AGE_RANGE", "MARITAL_STATUS", "FAMILY_SIZE", "PROFESSION", "EDUCATION_LEVEL")]
    #   demoDeets[["CUSTOMER_ID"]] <- as.integer(demoDeets[["CUSTOMER_ID"]])
    #   demoDeets[["FAMILY_SIZE"]] <- as.integer(demoDeets[["FAMILY_SIZE"]])
    #   demoDeets[["ADDRESS"]] <- paste(selection[,"ADDRESS_HOME_CITY"], selection[,"ADDRESS_HOME_STATE"], sep = ', ')
    #   demoDeets[,c("CUSTOMER_ID", "AGE_RANGE", "ADDRESS", "MARITAL_STATUS", "FAMILY_SIZE", "PROFESSION", "EDUCATION_LEVEL")]
    # }, bordered = TRUE, align = 'l')
    
    output$customerInfo <- renderUI({
      infoDets <- selection[,c("AGE_RANGE", "MARITAL_STATUS", "FAMILY_SIZE", "PROFESSION", "EDUCATION_LEVEL")]
      infoDets[["FAMILY_SIZE"]] <- as.integer(infoDets[["FAMILY_SIZE"]])
      infoDets[["ADDRESS"]] <- paste(selection[,"ADDRESS_HOME_CITY"], selection[,"ADDRESS_HOME_STATE"], sep = ', ')
      tags$ul( class = 'list-unstyled',
               tags$li(
                 tags$strong('Age: '), tags$span(class = "pull-right", infoDets[["AGE_RANGE"]], ' years old')
               ),
               tags$li(
                 tags$strong('Marital Status: '),tags$span(class = "pull-right", infoDets[["MARITAL_STATUS"]])
               ),
               tags$li(
                 tags$strong('Address: '), tags$span(class = "pull-right", infoDets[["ADDRESS"]])
               ),
               tags$li(
                 tags$strong('Profession: '), tags$span(class = "pull-right",infoDets[["PROFESSION"]])
               ),
               tags$li(
                 tags$strong('Level of Education: '), tags$span(class = "pull-right", infoDets[["EDUCATION_LEVEL"]])
               )
      )
    })
    
    
    # output$customerFinancesTable <- renderTable({
    #   finDeets <- selection[,c("ANNUAL_INCOME", "HOME_OWNER_INDICATOR", "MONTHLY_HOUSING_COST", "CREDIT_SCORE", "CREDIT_AUTHORITY_LEVEL")]
    #   finDeets[["ANNUAL_INCOME"]] <- dollar(finDeets[["ANNUAL_INCOME"]])
    #   finDeets[["MONTHLY_HOUSING_COST"]] <- dollar(finDeets[["MONTHLY_HOUSING_COST"]])
    #   finDeets
    # }, bordered = TRUE, align = 'l')
    
    output$customerFinancesInfo <- renderUI({
      customerfinDets <- selection[,c("ANNUAL_INCOME", "HOME_OWNER_INDICATOR", "MONTHLY_HOUSING_COST", "CREDIT_SCORE", "CREDIT_AUTHORITY_LEVEL")]
      customerfinDets[["ANNUAL_INCOME"]] <- dollar(customerfinDets[["ANNUAL_INCOME"]])
      customerfinDets[["MONTHLY_HOUSING_COST"]] <- dollar(customerfinDets[["MONTHLY_HOUSING_COST"]])
      tags$ul( class = 'list-unstyled',
               tags$li(
                 tags$strong('Annual income: '), tags$span(class="pull-right", customerfinDets[["ANNUAL_INCOME"]])
               ),
               tags$li(
                 tags$strong('Home Owner: '), tags$span(class="pull-right", if(customerfinDets[["HOME_OWNER_INDICATOR"]] == TRUE) { 'Yes'} else { 'No'})
               ),
               tags$li(
                 tags$strong('Monthly Housing: '), tags$span(class="pull-right", customerfinDets[["MONTHLY_HOUSING_COST"]])
               ),
               tags$li(
                 tags$strong('Credit Score: '), tags$span(class="pull-right", round(customerfinDets[["CREDIT_SCORE"]], 0))
               ),
               tags$li(
                 tags$strong('Credit Authority Level: '), tags$span(class="pull-right", customerfinDets[["CREDIT_AUTHORITY_LEVEL"]])
               )
               
      )
    })
    
    customerAccounts <- account[account$PRIMARY_CUSTOMER_ID == sessionVars$selectedClientId,][1,]
    
    output$customerAccountsTable <- renderTable({
      customerAccountsInfo <- customerAccounts[,c("ACCOUNT_ID", "ACCOUNT_TYPE", "PRODUCT_ID", "BASE_CURRENCY", "INVESTMENT_OBJECTIVE", "LIFE_CYCLE_STATUS", "RISK_TOLERANCE", "TAX_ADVANTAGE_INDICATOR")]
      for(name in names(customerAccountsInfo)){
        colnames(customerAccountsInfo)[colnames(customerAccountsInfo)==name] <- str_replace_all(name, '_',' ')
      }
      customerAccountsInfo
      },
      bordered = FALSE, align = 'l')
    
    customerAccountSummaries <- account_summary[account_summary$ACCOUNT_ID %in% customerAccounts$ACCOUNT_ID,][1,]
    
    output$customerAccountSummaryTable <- renderTable({
      accountDeets <- customerAccountSummaries[,c("CLOSING_BALANCE", "AMOUNT_OF_DEPOSITS", "AMOUNT_OF_INTEREST_EARNED", "NUMBER_OF_BUY_TRADES", "AMOUNT_OF_BUY_TRADES", "AMOUNT_OF_MARKET_CHANGE")] %>%
        mutate_at(vars(contains("AMOUNT_")), dollar) %>%
        mutate_at(vars(contains("_BALANCE")), dollar) %>%
        mutate_at(vars(contains("NUMBER_")), as.integer)
      for(name in names(accountDeets)){
        colnames(accountDeets)[colnames(accountDeets)==name] <- str_replace_all(name, '_',' ')
      }
      accountDeets
    }, bordered = FALSE, align = 'l')
    
    # Reset scoring
    removeUI(selector = "#scoringResponse > *", multiple = TRUE)
    shinyjs::hide(id = "scoringResponse")
    shinyjs::show(id = "scoreBtnSection")
    output$scoringError <- renderText('')
    sessionVars$pipelineInput <- list(cust_id = sessionVars$selectedClientId, sc_end_date = '2018-09-30')
    output$pipelineInput <- renderText(toJSON(sessionVars$pipelineInput, indent = 2))
  })
  
  # Set default hostname for CP4D API
  observeEvent(session$clientData$url_hostname, {
    updateTextInput(session, "hostname", value = session$clientData$url_hostname)
  })
  
  # Enable buttons when inputs are provided
  observe({
    toggleState("authBtn", nchar(input$hostname) > 0 && nchar(input$username) > 0 && nchar(input$password) > 0)
    toggleState("scoreBtn", nchar(input$endpoint) > 0 && nchar(input$token) > 0 && length(input$allCustomers_rows_selected) > 0)
  })
  
  # Handle CP4D API authentication button
  observeEvent(input$authBtn, {
    shinyjs::disable("authBtn")
    
    tryCatch({
      deploymentsResp = collectDeployments(input$hostname, input$username, input$password, "Customer-Attrition-Prediction-Scoring-Function-Deployment")
      serverVariables$deployments <- deploymentsResp$deployments
      serverVariables$token = deploymentsResp$token
    }, warning = function(w) {
      output$authError <- renderText(w$message)
    }, error = function(e) {
      output$authError <- renderText(e$message)
    })
    
    shinyjs::enable("authBtn")

    if(length(serverVariables$deployments) > 0) {
      updateSelectInput(session, "deploymentSelector", choices = names(serverVariables$deployments))
      shinyjs::hide(id = "authPanel")
      shinyjs::show(id = "deploymentPanel")
    }
  })
  
  # Handle model deployment dropdown switching
  observeEvent(input$deploymentSelector, {
    selectedDeployment <- serverVariables$deployments[[input$deploymentSelector]]
    output$deployment_guid <- renderText(selectedDeployment$guid)
    output$space_name <- renderText(selectedDeployment$space_name)
   
    output$scoring_url <- renderText(selectedDeployment$scoring_url)
    toggleState("scoreBtn", nchar(selectedDeployment$scoring_url) > 0 && nchar(serverVariables$token) > 0)
  })
  
  # Handle model deployment scoring button
  observeEvent(input$scoreBtn, {
    shinyjs::disable("scoreBtn")
    
    selectedDeployment <- serverVariables$deployments[[input$deploymentSelector]]
    

    payload = list(
      fields=list("CUSTOMER_ID","sc_end_date"),
      values = list(list(sessionVars$selectedClientId,'2018-09-30'))
    )
    
    print(payload)
    print(serverVariables$token)
    response <- scoreModelDeployment(selectedDeployment$scoring_url, payload, serverVariables$token)

    if(length(response$error) > 0) {
      output$scoringError <- renderText(toString(response$error))
    }
    else if(length(response$predictions) > 0) {
      shinyjs::hide(id = "scoreBtnSection")
      shinyjs::show(id = "scoringResponse")
      
      result <- response$predictions[[1]]$values$predictions[[1]]
      # has_explain <- length(result$explain_plot_html) > 0
      
      # explain_panel <- div()
      # if(has_explain) {
      #   explain_panel <- div(
      #     h4("Highest Impact Features: "),
      #     br(),
      #     p(
      #       tableOutput("explainTable")
      #     ),
      #     br(),
      #     h4("Explanation Plot: "),
      #     HTML(result$explain_plot_html)
      #   )
      # }
      
      insertUI(
        selector = "#scoringResponse",
        where = "beforeEnd",
        ui = panel(
          h3("Customer Attrition Prediction:"),
          p(
            plotOutput("probPlot", width = "600px", height = "300px")
          )
          # explain_panel
        )
      )
      
      # if(has_explain) {
      #   # send responseInserted message
      #   session$sendCustomMessage('responseInserted', 
      #                             list(
      #                               id=result$explain_plot_elem_id,
      #                               data=fromJSON(result$explain_plot_data))
      #   )
      #   
      #   # render high impact features table
      #   vertical <- t(data.frame(result$explain))
      #   Impact <- vertical[order(vertical, decreasing = TRUE),]
      #   df <- data.frame(Impact)
      #   dispTable <- tibble::rownames_to_column(df, "Feature")
      #   for(feature_name in dispTable[["Feature"]]) {
      #     feature_newName <- str_replace_all(feature_name, "_", " ")
      #     feature_newName <- str_replace_all(feature_newName, '[.]', " ")
      #     dispTable[["Feature"]][dispTable[["Feature"]] == feature_name] <- feature_newName
      #   }
      #   output$explainTable <- renderTable(dispTable, bordered = TRUE)
      # }
      
      # generate probability pie
      probDF <- data.frame(t(data.frame(result$values[[1]][[2]])))
      colnames(probDF) <- "Probability"
      row.names(probDF) <- c("FALSE", "TRUE")
      probDF <- tibble::rownames_to_column(probDF, "Prediction")
      probDF <- probDF %>%
        mutate(percentage = paste0(round(100 * Probability, 1), "%")) %>%
        mutate(hover_text = paste0(Prediction, ": ", percentage))
      
      probPlot <- ggplot(probDF, aes(y = Probability, fill = Prediction)) +
        geom_bar(
          aes(x = 1),
          width = 0.4,
          stat = "identity",
          show.legend = TRUE
        ) + 
        annotate("text", x = 0, y = 0, size = 12,
                 label = probDF[["percentage"]][probDF[["Prediction"]] == "TRUE"]
                ) +
        coord_polar(theta = "y") +
        theme_void() +
        theme(legend.title=element_text(size=22),
              legend.text=element_text(size=16)) +
        guides(fill = guide_legend(reverse=TRUE))
      
      output$probPlot <- renderPlot(probPlot, width="auto", height="auto")
      
    } else {
      output$scoringError <- renderText(response)
    }
    
    shinyjs::enable("scoreBtn")
  })
  
  observeEvent(input$reauthenticateBtn, {
    shinyjs::show(id = "scoreBtnSection")
    removeUI(selector = "#scoringResponse > *", multiple = TRUE)
    shinyjs::show(id = "authPanel")
    shinyjs::hide(id = "deploymentPanel")
  })
}

File: assets/rstudio/Dash-customer-attrition-prediction-analytics-dashboard/clientPanel.py
from lib.icp4d_api import cp4d_api_function
from lib.load_data import load_data
import dash
import dash_html_components as html
import dash_core_components as dcc
from dash.dependencies import Input, Output, State

# Import necessary libraries
import pandas as pd





# Load the data
data = load_data()

# Define the layout
client_layout = html.Div(
    [
        dcc.Tabs(
            [
                dcc.Tab(label="Client View", value="clientPanel", children=[
                    html.Br(),
                    html.Div(
                        className="pull-left",
                        children=[
                            html.Div(id="customerImage"),
                            html.H2(className="text-center",
                                    id="customerName"),
                        ],
                        style={"width": "25%"},
                    ),
                    html.Div(
                        children=[
                            html.H4("Personal Information",
                                    className="text-center"),
                            html.Hr(),
                            html.Div(id="customerInfo")
                        ]
                    ),
                    html.Div(
                        className="pull-right",
                        children=[
                            html.H4("Financial Profile",
                                    className="text-center"),
                            html.Hr(),
                            html.Div(id="customerFinancesInfo")
                        ],
                        style={"width": "25%", "float": "right"}
                    )
                ]),
                dcc.Tab(label="Account Details", value="accountDetails", children=[
                    html.H3("Account Details"),
                    html.Br(),
                    html.Table(id="customerAccountsTable"),
                    html.Br(),
                    html.Table(id="customerAccountSummaryTable")
                ]),
                dcc.Tab(label="Attrition Prediction", value="attritionPrediction", children=[
                    html.H3("Attrition Prediction"),
                    html.Br(),
                    html.Div(
                        [
                            html.Div(
                                id="authPanel",
                                children=[
                                    html.H4(
                                        "Connect to Cloud Pak for Data API"),
                                    dcc.Input(id="hostname",
                                              placeholder="CPD Hostname"),
                                    dcc.Input(id="username",
                                              placeholder="CPD Username"),
                                    dcc.Input(
                                        id="password", placeholder="CPD Password", type="password"),
                                    html.Button("Authenticate API", id="authBtn", n_clicks=0,
                                                className="btn-primary btn-lg btn-block", style={"max-width": "300px"}),
                                    html.Div(id="authError", style={
                                             "color": "red"})
                                ],
                                style={"max-width": "360px"}
                            ),
                            html.Div(
                                id="deploymentPanel",
                                children=[
                                    html.Div(
                                        children=[
                                            html.H4(
                                                "Model Scoring Pipeline Deployment"),
                                            dcc.Dropdown(
                                                id="deploymentSelector",
                                                options=[],
                                                value="",
                                                placeholder="Deployment:"
                                            ),
                                            html.P([
                                                html.Strong("Space Name: "),
                                                html.Span(
                                                    id="space_name", className="pull-right", style={"word-wrap": "break-word"})
                                            ]),
                                            html.P([
                                                html.Strong("GUID: "),
                                                html.Span(
                                                    id="deployment_guid", className="pull-right", style={"word-wrap": "break-word"})
                                            ]),
                                            html.P([
                                                html.Strong(
                                                    "Scoring Endpoint: "),
                                                html.Span(
                                                    id="scoring_url", className="pull-right", style={"word-wrap": "break-word"})
                                            ])
                                        ],
                                        style={"max-width": "360px"}
                                    ),
                                    html.Div(
                                        children=[
                                            html.Button(
                                                "Re-Authenticate", id="reauthenticateBtn", className="btn-primary btn-lg btn-block")
                                        ]
                                    )
                                ]
                            ),
                            html.Div(
                                id="scoreBtnSection",
                                children=[
                                    html.Div(
                                        className="pull-left",
                                        children=[
                                            html.Button("Predict Attrition", id="scoreBtn", n_clicks=0,
                                                        className="btn-primary btn-lg btn-block", style={"max-width": "300px"}),
                                            html.Br(),
                                            html.H4("Input JSON:"),
                                            html.Pre(id="pipelineInput"),
                                            html.Div(id="scoringError",
                                                     style={"color": "red"})
                                        ]
                                    ),
                                    html.Div(
                                        style={"width": "70%",
                                               "float": "right"},
                                        id="scoringResponse"
                                    )
                                ],
                                style={"display": "none"}
                            )
                        ]
                    )
                ])
            ],
            id="tabs",
            value="clientPanel"
        ),
    ]
)

# Define the callbacks


@app.callback(
    Output("customerName", "children"),
    Output("customerImage", "children"),
    Output("customerInfo", "children"),
    Output("customerFinancesInfo", "children"),
    Output("customerAccountsTable", "children"),
    Output("customerAccountSummaryTable", "children"),
    Output("pipelineInput", "children"),
    [Input("tabs", "value")],
    [State("tabs", "value")]
)
def update_client_view(selected_tab, current_tab):
    if selected_tab == "clientPanel":
        client = data["client"]

        customer_name = client["name"]
        customer_image = html.Img(
            src="profiles/" + client["image"],
            style={"display": "block", "margin-left": "auto",
                   "margin-right": "auto"},
            width=150,
            height=150
        )

        customer_info = html.Ul(
            className="list-unstyled",
            children=[
                html.Li([
                    html.Strong("Age: "),
                    html.Span(client["age_range"], className="pull-right",),
                    " years old"
                ]),
                html.Li([html.Strong("Marital Status: "), html.Span(
                    client["marital_status"], className="pull-right")]),
                html.Li([html.Strong("Address: "), html.Span(
                    client["address"], className="pull-right")]),
                html.Li([html.Strong("Profession: "), html.Span(
                    client["profession"], className="pull-right")]),
                html.Li([html.Strong("Level of Education: "), html.Span(
                    client["education_level"], className="pull-right")])
            ]
        )

        customer_finances_info = html.Ul(
            className="list-unstyled",
            children=[
                html.Li([html.Strong("Annual income: "), html.Span(
                    client["annual_income"], className="pull-right")]),
                html.Li([html.Strong("Home Owner: "), html.Span(
                    "Yes" if client["home_owner_indicator"] else "No", className="pull-right")]),
                html.Li([html.Strong("Monthly Housing: "), html.Span(
                    client["monthly_housing_cost"], className="pull-right")]),
                html.Li([html.Strong("Credit Score: "), html.Span(
                    round(client["credit_score"], 0), className="pull-right")]),
                html.Li([html.Strong("Credit Authority Level: "), html.Span(
                    client["credit_authority_level"], className="pull-right")])
            ]
        )

        customer_accounts_table = html.Table(
            className="table",
            children=[
                html.Thead(
                    html.Tr(
                        [html.Th(col) for col in ["Account ID", "Account Type", "Product ID", "Base Currency",
                                                  "Investment Objective", "Life Cycle Status", "Risk Tolerance", "Tax Advantage Indicator"]]
                    )
                ),
                html.Tbody(
                    [
                        html.Tr(
                            [html.Td(customer["account"][col]) for col in ["account_id", "account_type", "product_id", "base_currency",
                                                                           "investment_objective", "life_cycle_status", "risk_tolerance", "tax_advantage_indicator"]]
                        )
                    ]
                )
            ]
        )

        customer_account_summary_table = html.Table(
            className="table",
            children=[
                html.Thead(
                    html.Tr(
                        [html.Th(col) for col in ["Closing Balance", "Amount of Deposits", "Amount of Interest Earned",
                                                  "Number of Buy Trades", "Amount of Buy Trades", "Amount of Market Change"]]
                    )
                ),
                html.Tbody(
                    [
                        html.Tr(
                            [html.Td(customer["account_summary"][col]) for col in ["closing_balance", "amount_of_deposits",
                                                                                   "amount_of_interest_earned", "number_of_buy_trades", "amount_of_buy_trades", "amount_of_market_change"]]
                        )
                    ]
                )
            ]
        )

        pipeline_input = html.Pre(
            json.dumps(
                {"cust_id": client["id"], "sc_end_date": "2018-09-30"}, indent=2),
            id="pipelineInput"
        )

        return customer_name, customer_image, customer_info, customer_finances_info, customer_accounts_table, customer_account_summary_table, pipeline_input

    return "", "", "", "", "", "", ""


@app.callback(
    Output("space_name", "children"),
    Output("deployment_guid", "children"),
    Output("scoring_url", "children"),
    Output("scoreBtnSection", "style"),
    Output("authError", "children"),
    Output("authPanel", "style"),
    Output("deploymentSelector", "options"),
    Output("deploymentSelector", "value"),
    [Input("authBtn", "n_clicks")],
    [State("hostname", "value"), State("username", "value"), State(
        "password", "value"), State("deploymentSelector", "value")]
)
def authenticate_api(n_clicks, hostname, username, password, deployment_selector):
    if n_clicks == 0:
        return "", "", "", {"display": "none"}, "", "", [], ""

    # Authenticate API here
    try:
        deployments_resp = cp4d_api_function(
            hostname, username, password, "Customer-Attrition-Prediction-Scoring-Function-Deployment")
        deployments = deployments_resp["deployments"]
        token = deployments_resp["token"]
    except Exception as e:
        return "", "", "", {"display": "none"}, str(e), "", [], ""

    deployment_options = [{"label": deployment, "value": deployment}
                          for deployment in deployments]
    deployment_value = deployment_selector if deployment_selector in deployments else ""

    return deployments[deployment_value]["space_name"], deployments[deployment_value]["guid"], deployments[deployment_value]["scoring_url"], {}, "", {"max-width": "360px"}, deployment_options, deployment_value


@app.callback(
    Output("scoreBtnSection", "style"),
    [Input("reauthenticateBtn", "n_clicks")]
)
def reauthenticate_api(n_clicks):
    if n_clicks > 0:
        return {"display": "none"}

    return {}


@app.callback(
    Output("scoreBtnSection", "style"),
    [Input("deploymentSelector", "value"), Input("token", "value"),
     Input("allCustomers_rows_selected", "children")]
)
def enable_score_btn(deployment_selector, token, all_customers_rows_selected):
    if not deployment_selector or not token or not all_customers_rows_selected:
        return {"display": "none"}

    return {}


@app.callback(
    Output("scoringResponse", "children"),
    Output("scoringError", "children"),
    [Input("scoreBtn", "n_clicks"), Input("deploymentSelector", "value"), Input(
        "token", "value"), Input("allCustomers_rows_selected", "children")]
)
def score_model_deployment(n_clicks, deployment_selector, token, all_customers_rows_selected):
    if n_clicks == 0:
        return [], ""

    selected_deployment = deployments[deployment_selector]

    payload = {
        "fields": ["CUSTOMER_ID", "sc_end_date"],
        "values": [[all_customers_rows_selected, "2018-09-30"]]
    }

    try:
        response = scoreModelDeployment(
            selected_deployment["scoring_url"], payload, token)

        if "error" in response:
            return [], response["error"]
        elif "predictions" in response:
            result = response["predictions"][0]["values"]["predictions"][0]

            # Generate probability pie
            prob_values = result["values"][1]
            prob_df = pd.DataFrame(prob_values, columns=[
                                   "Probability"]).transpose()
            prob_df.columns = ["Prediction"]
            prob_df["Percentage"] = (
                prob_df["Probability"] * 100).round(1).astype(str) + "%"
            prob_df["Hover Text"] = prob_df["Prediction"] + \
                ": " + prob_df["Percentage"]

            prob_plot = get_prob_plot(prob_df)

            return [
                html.H3("Customer Attrition Prediction:"),
                html.P(prob_plot),
                html.Pre(json.dumps(result, indent=2), id="pipelineInput"),
            ], ""
    except Exception as e:
        return [], str(e)


if __name__ == "__main__":
    app.run_server(debug=True)
